ARROW-13806: [C++][Python] Add support for new Interval Type

emkornfield opened the pull request:
- Refactored ObjectWriter helpers from arrow_to_pandas, so they can be
  used for plain python types as well (generalized the lowest level so
  it can work on both PyObject** and an adapter for PyList.

- Add DateOffset to static pandas imports

- Tried to start laying out code in a way to use C++ for Array.to_pylist
  (feel free to comment).

Support importing from timeinterval, relativedelta and DateOffset types
(this is actually mostly duck types, the one complication is that
 relativedelta has a property weeks that is automatically calculated, so
 some type checking is necessary).

Open questions:
 - Should we be more strict on duck typing imports? I chose generalism
   over performance here (rechecking non-present attributes, etc)?
 - Is the new arrow_to_python.h desirable (I think this can be easily
   extended for other types)?
 - My python is rusty and Python C-API even more so, please don't assume
   I know exactly what I'm doing :)

emkornfield commented:
CC @tswast @jorisvandenbossche

pitrou commented on cpp/src/arrow/python/CMakeLists.txt:
```diff
@@ -28,6 +28,7 @@ add_dependencies(arrow_python-all arrow_python arrow_python-tests)
 
 set(ARROW_PYTHON_SRCS
     arrow_to_pandas.cc
+    arrow_to_python.cc
```
Looks like you forgot to add this file?

emkornfield replied:
added.

pitrou commented on cpp/src/arrow/python/datetime.cc:
```diff
@@ -71,6 +71,19 @@ bool MatchFixedOffset(const std::string& tz, util::string_view* sign,
+    {nullptr, nullptr}};
+
+static PyStructSequence_Desc MonthDayNanoTupleDesc = {
+    "MonthDayNanoTuple", "A interval consistent of months, days and nanoseconds.",
```
What does "consistent" mean here?

emkornfield replied:
meant to say consisting

pitrou commented on cpp/src/arrow/python/datetime.cc:
```diff
@@ -270,6 +283,18 @@ static inline Status PyDate_convert_int(int64_t val, const DateUnit unit, int64_
+  return (PyObject*)&MonthDayNanoTupleType;
+}
+
+PyTypeObject* BorrowMonthDayNanoTupleType() { return &MonthDayNanoTupleType; }
```
This doesn't seem used?

emkornfield replied:
yep, left over from a prior design. removed.

pitrou commented on cpp/src/arrow/python/datetime.cc:
```diff
@@ -450,6 +475,19 @@ Result<std::string> TzinfoToString(PyObject* tzinfo) {
   return PyTZInfo_utcoffset_hhmm(tzinfo);
 }
 
+Result<PyObject*> MonthDayNanoIntervalToNamedTuple(
```
Since NULL is return an error, you don't need to wrap the return value in a `Result<>`.

emkornfield replied:
done.

emkornfield replied:
There are a few other places that return Result<PyObject*> is there any guidance on when to use that and when to use pure PyObject*?

pitrou commented on cpp/src/arrow/python/python_to_arrow.cc:
```diff
@@ -68,6 +69,95 @@ using internal::MakeConverter;
 
 namespace py {
 
+enum class MonthDayNanoField { kMonths, kWeeksAndDays, kDaysOnly, kNanoseconds };
```
Since we are at it, can you perhaps enclose all private APIs in the anonymous namespace?

emkornfield replied:
i had these in an anonymous namespace originally and got obscure linking errors.  I can try it again to see if I get different results.

emkornfield replied:
After moving code around i was able to move the new ones to private.  I'll play around with adding some more.

pitrou commented on python/pyarrow/tests/test_array.py:
```diff
@@ -2162,6 +2163,101 @@ def test_array_from_numpy_ascii():
     assert arrow_arr.equals(expected)
 
 
+@pytest.mark.nopandas
```
"nopandas" is for tests that fail if Pandas is installed, I think.

emkornfield replied:
removed.  before this was testing the different return of to_pylist.

pitrou commented on python/pyarrow/tests/test_array.py:
```diff
@@ -2162,6 +2163,101 @@ def test_array_from_numpy_ascii():
+    assert arr.type == pa.month_day_nano_interval()
+    expected_list = [
+        None,
+        pa.MonthDayNanoTuple([0, 8,
```
Can you test with a non-zero number of months?

emkornfield replied:
I might be missing something but it doesn't appear  timedelta has a months or years field, which would be what is required to have a non-zero months fields (otherwise days is sufficient).

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -514,6 +514,32 @@ cdef class DurationScalar(Scalar):
+    def as_py(self):
+        """
+        Return this value as a Pandas DateOffset instance if Pandas is present
+        otherwise as a named tuple containing months days and nanoseconds.
```
I'm a bit concerned that this makes user code fragile: depending on whether Pandas is installed or not, you'll get an object with a different API.

Instead, perhaps always return a `pa.MonthDayNano`, but add a `MonthDayNano.to_pandas()` method to produce a Pandas DateOffset?

emkornfield replied:
I kind of agree, I modeled this after  [Timestamps](https://github.com/apache/arrow/blob/master/python/pyarrow/scalar.pxi#L382)  which also changes behavior.  Happy to simplify this though if that is desired (along with corresponding to_pylist()) logic.

pitrou replied:
Hmm, it's a bit unfortunate for timestamps, but the decision there is understandable: returning a `datetime.datetime` loses precision. Also, `pandas.Timestamp` is probably API-compatible with `datetime.datetime`.

You don't lose any precision by returning a `MonthDayNano`.

westonpace replied:
I'm going to agree with @pitrou .  With pyarrow we even have cases where users aren't even really aware they are using pyarrow.  I could see a circumstance where they install their application on some new machine or environment and suddenly start getting errors.  Maybe someday we can add a `use_pandas_types` option to `as_py`, `to_pylist`, and `to_pydict`.  That way we can error if the expected library is not available instead of silently changing the type.

emkornfield replied:
Going to change.  This actually simplifies the code a lot.

jorisvandenbossche replied:
Yes, I also agree returning a consistent type here is best (for Timestamp there is the complication of datetime.datetime not supporting nanoseconds ..). As Weston mentions, we can see later if we want to add some option to get a more pandas native object.

pitrou commented on python/pyarrow/lib.pyx:
```diff
@@ -37,6 +37,10 @@ arrow_init_numpy()
 
+cdef libarrow.ArrowToPython ARROW_TO_PYTHON
+
+MonthDayNanoTuple = NewMonthDayNanoTupleType()
```
Nit, but I would simply call it `MonthDayNano`.

emkornfield replied:
Done.

pitrou commented on python/pyarrow/lib.pyx:
```diff
@@ -37,6 +37,10 @@ arrow_init_numpy()
 # (used from some of our C++ code, see e.g. ARROW-5260)
 import_pyarrow()
 
+cdef libarrow.ArrowToPython ARROW_TO_PYTHON
```
This seems rather cryptic. What does this do? Is this a globally defined converter?
(is initialization costly?)

emkornfield replied:
yes, globally defined converter.  This will likely change based on the outcome of simplying "as_py" but before it required getting references to Pandas types.  In the long run we likely need to import a few others (timestamp, timedelta, decimal, etc).  It seems like for Pandas we used C++ static variables.  It felt cleaner to me make it an explicit object.  Again don't feel too strongly here and if we eliminate the immediate need to share Pandas conversion logic this could go away.  Do you have a preference on the pattern here?

pitrou replied:
I'm ok with having a wrapper object, however we shouldn't import Pandas unconditionally at startup as that's quite costly.

emkornfield replied:
It no longer needs to do any imports for this PR (pandas objects are left in pandas related classes).  It was doing the import lazily on first method calls.

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -1508,6 +1508,30 @@ cdef class DurationArray(NumericArray):
+
+        is installed the objects will be
+        pd.tseries.offsets.DateOffset objects.  Otherwise they are
+        pyarrow.MonthDayNanoTuple objects.
```
Same remark as for the corresponding scalar class. Also, it seems the docstring is a bit garbled?

emkornfield replied:
Should be fixed now (once I push latest set of commits.

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -1508,6 +1508,30 @@ cdef class DurationArray(NumericArray):
     Concrete class for Arrow arrays of duration data type.
     """
 
+cdef class MonthDayNanoIntervalArray(Array):
```
Nit, but please let's remain consistent about spacing between classes.

emkornfield replied:
done

pitrou commented on cpp/src/arrow/python/python_to_arrow.cc:
```diff
@@ -68,6 +69,95 @@ using internal::MakeConverter;
+        return Status::Invalid("Overflow on: ", (attr - 1)->name);
+      }
+      if (PyObject_HasAttrString(obj, attr->name)) {
+        OwnedRef field_value(PyObject_GetAttrString(obj, attr->name));
```
If you want to avoid a double hash lookup, you can simply call `PyObject_GetAttrString` and catch the `AttributeError` on failure.

emkornfield replied:
Good point. Done.

pitrou commented on cpp/src/arrow/python/python_to_arrow.cc:
```diff
@@ -68,6 +69,95 @@ using internal::MakeConverter;
+
+template <MonthDayNanoField field>
+struct PopulateMonthDayNano {
+  using Traits = MonthDayNanoTraits<field>;
```
Add a `using c_type = ...` as well?

emkornfield replied:
done

pitrou commented on cpp/src/arrow/python/python_to_arrow.cc:
```diff
@@ -68,6 +69,95 @@ using internal::MakeConverter;
+        OwnedRef field_value(PyObject_GetAttrString(obj, attr->name));
+        RETURN_IF_PYERROR();
+        *found_attrs = true;
+        if (field_value.obj() == Py_None) {
```
What does `None` mean in this context?

emkornfield replied:
i thought it might have been a valid value.  remove this check.

emkornfield commented:
@pitrou thanks for the feedback, I'll address the ones I haven't commented on.  I think the most substantive one is whether return types should be different with or without Pandas.  Happy to go with whatever you and @jorisvandenbossche think is best.

westonpace commented on python/pyarrow/tests/test_array.py:
```diff
@@ -2162,6 +2163,101 @@ def test_array_from_numpy_ascii():
+        None,
+        relativedelta(years=1, months=1,
+                      days=1, seconds=1, microseconds=1,
+                      minutes=1, hours=1, weeks=1, leapdays=1)]
+    # Note leapdays are ignored.
```
Nit: Can we mention this in at least one place in a docstring.

emkornfield replied:
added to pa.array(...) doc string.

westonpace commented on python/pyarrow/types.pxi:
```diff
@@ -2154,6 +2154,14 @@ def duration(unit):
     return out
 
 
+def month_day_nano_interval():
+    """
+    Create instance of a interval representing the time between two calendar
+    instances represented as a triple of months, days and nanoseconds.
+    """
+    return primitive_type(_Type_INTERVAL_MONTH_DAY_NANO)
```
Nit: I would avoid "the time between two calendar instances" because a duration/Joda-interval is traditionally defined as "the time between two instants".  I haven't seen intervals / periods defined in this way before (possibly because calendars could change).

Also, instead of `a triple of months` (which might make a python user think 3-tuple) could we say "represented by values for months, days, and nanoseconds"?  Or, if we want to be real precise, "represented by a signed 32 bit integer of months, a signed 32 bit integer of days, and a signed 64 bit integer of nanoseconds".

emkornfield replied:
copied language from type.h

westonpace commented on python/pyarrow/scalar.pxi:
```diff
@@ -514,6 +514,32 @@ cdef class DurationScalar(Scalar):
+    @property
+    def value(self):
+        """
+        Returns this value pyarrow.MonthDayNanoTuple.
```
```suggestion
        Returns this value as a pyarrow.MonthDayNanoTuple.
```

emkornfield commented on cpp/src/arrow/python/arrow_to_python.h:
```diff
@@ -0,0 +1,161 @@
+inline Status WriteArrayObjects(const ArrayType& arr, WriteValue&& write_func,
+                                Assigner out_values) {
+  // TODO(ARROW-12976): Use visitor here?
+  const bool has_nulls = arr.null_count() > 0;
+  for (int64_t i = 0; i < arr.length(); ++i) {
+    if (has_nulls && arr.IsNull(i)) {
+      Py_INCREF(Py_None);
+      *out_values = Py_None;
+    } else {
+      RETURN_NOT_OK(write_func(arr.GetView(i), out_values));
+    }
+    ++out_values;
+  }
```
this was moved from existing code base, I don't know but would prefer to handle the TODO above use Visitor in general.

emkornfield commented:
@pitrou @westonpace thanks for the reviews I think I've addressed all of the comments.

pitrou commented on cpp/src/arrow/python/arrow_to_pandas.cc:
```diff
@@ -1097,6 +1081,40 @@ struct ObjectWriterVisitor {
+      DCHECK(internal::BorrowPandasDataOffsetType() != nullptr);
+      // TimeDelta objects do not add nanoseconds component to timestamp.
+      // so convert microseconds and remainder to preserve data
+      // but give users more expected results.
```
Code looks ok, but I don't understand this comment. What do you mean with "do not add nanoseconds component to timestamp"?

emkornfield replied:
I'll try to clarify but there are actually two bugs:
https://github.com/pandas-dev/pandas/issues/36589
https://github.com/pandas-dev/pandas/issues/43892

This is meant to mitigate the latter.  I'll link the second issue here.

CC @tswast as an FYI

emkornfield replied:
Rephrased (had the wrong time should have been DateOffset) and referenced the bug.

pitrou replied:
Ah, thank you.

pitrou commented on cpp/src/arrow/python/arrow_to_python.h:
```diff
@@ -0,0 +1,59 @@
+// specific language governing permissions and limitations
+// under the License.
+
+// Functions for converting between pandas's NumPy-based data representation
```
Do these functions have anything to do with Pandas?

emkornfield replied:
I'll update the comment, I removed the numpy/pandas stuff for now, not sure if it will be required here or of pandas/numpy can continue to live in arrow_to_pandas as part of ARROW-12976

emkornfield replied:
removed for now.

pitrou commented on cpp/src/arrow/python/arrow_to_python.h:
```diff
@@ -0,0 +1,59 @@
+
+namespace py {
+
+/// \brief Utility class for converting Arrow to Python obects.  A class instead
```
Trailing unfinished sentence here?

emkornfield replied:
yeah, covered below.

pitrou commented on cpp/src/arrow/python/arrow_to_python.h:
```diff
@@ -0,0 +1,59 @@
+  // offset from the unix epoch.
+  ///
+  /// N.B. This has limited type support.  ARROW-12976 tracks full implementation.
+  Result<PyObject*> ToPrimitive(const Scalar& scalar);
```
Why isn't this called "ToPyObject"? Is it only for primitive types?

emkornfield replied:
See comment above for explanation, but I think ToPyObject isn't specific enough when it comes to ARROW-12976

emkornfield replied:
After thinking about this some more, I don't think it is worth porting over the the "value" property code, only the as_py code.  I'm going to rename this.

pitrou commented on cpp/src/arrow/python/arrow_to_python.h:
```diff
@@ -0,0 +1,59 @@
+  /// representation.
+  ///
+  /// For instance timestamp would be translated to a integer representing an
+  // offset from the unix epoch.
```
Would it? I thought a timestamp would be converted to a datetime.

pitrou commented on cpp/src/arrow/python/arrow_to_python.h:
```diff
@@ -0,0 +1,59 @@
+ public:
+  /// \brief Converts the given Array to a PyList object. Returns NULL if there
+  /// is an error converting the Array. The list elements are the same ones
+  /// generated via ToLogical()
```
What is ToLogical?

pitrou commented on cpp/src/arrow/python/arrow_to_python.cc:
```diff
@@ -0,0 +1,107 @@
+}  // namespace
+
+Result<PyObject*> ArrowToPython::ToPyList(const Array& array) {
+  RETURN_NOT_OK(CheckInterval(*array.type()));
```
So these methods only work for month_day_nano_interval? That seems like a weird API choice.

emkornfield replied:
The intent was to generalize them for other types (see TODO JIRAs in the header for supporting more types).

pitrou replied:
I see. I'd rather see a comprehensive refactor in another PR than a stub like this, which may end up abandoned.

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -163,6 +163,12 @@ def array(object obj, type=None, mask=None, size=None, from_pandas=None,
     representation). Timezone-naive data will be implicitly interpreted as
     UTC.
 
+    Pandas's DateOffsets and dateutil.relativedelta.relativedetla are by
```
"delta"

emkornfield replied:
fixed.

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -163,6 +163,12 @@ def array(object obj, type=None, mask=None, size=None, from_pandas=None,
+    Pandas's DateOffsets and dateutil.relativedelta.relativedetla are by
+    default converted as MonthDayNanoIntervalArray. relativedelta leapday's
+    are ignored as are all absolute fields on both objects. datetime.timedelta
+    also be converted to MonthDayNanoIntervalArray but require passing
```
"also be"?

emkornfield replied:
add "can"

pitrou commented on python/pyarrow/tests/test_scalars.py:
```diff
@@ -364,6 +367,14 @@ def test_duration_nanos_nopandas():
 
 
+def test_month_day_nano_interval():
+    triple = pa.MonthDayNano([3600, 3600, 3600])
```
Please use different values for components here.

emkornfield replied:
done

pitrou reviewed (changes requested):
I find it rather unwelcome that this adds an abstraction ("ArrowToPython") that takes care of a single type. I'm not sure if that refactor is potentially useful, but we should have a single facility for conversion to Python objects, not a bunch of unrelated ones.

emkornfield commented:
> I find it rather unwelcome that this adds an abstraction ("ArrowToPython") that takes care of a single type. I'm not sure if that refactor is potentially useful, but we should have a single facility for conversion to Python objects, not a bunch of unrelated ones.

Mentioned this elsewhere but I was trying to prep layout for ARROW-12976 is this is undesirable, I can make these simple functions for now.

jorisvandenbossche commented on python/pyarrow/array.pxi:
```diff
@@ -163,6 +163,12 @@ def array(object obj, type=None, mask=None, size=None, from_pandas=None,
+    default converted as MonthDayNanoIntervalArray. relativedelta leapday's
+    are ignored as are all absolute fields on both objects. datetime.timedelta
+    can also be converted to MonthDayNanoIntervalArray but require passing
+    MonthDayIntervalType explicitly.
```
```suggestion
    MonthDayNanoIntervalType explicitly.
```

jorisvandenbossche commented on cpp/src/arrow/python/helpers.cc:
```diff
@@ -321,6 +323,14 @@ void InitPandasStaticData() {
 
+  // Import DateOffset type
+  OwnedRef offsets;
+  if (internal::ImportModule("pandas.tseries.offsets", &offsets).ok()) {
```
It seems that DateOffset if available in the top-level pandas namespace, in which case I would import it from there

jorisvandenbossche commented on cpp/src/arrow/python/datetime.h:
```diff
@@ -178,6 +182,17 @@ Result<PyObject*> StringToTzinfo(const std::string& tz);
 ARROW_PYTHON_EXPORT
 Result<std::string> TzinfoToString(PyObject* pytzinfo);
 
+/// Converts MonthDayNano to a python dictionary.
```
```suggestion
/// Converts MonthDayNano to a python namedtuple.
```

jorisvandenbossche commented on cpp/src/arrow/python/datetime.h:
```diff
@@ -178,6 +182,17 @@ Result<PyObject*> StringToTzinfo(const std::string& tz);
 
+/// Converts MonthDayNano to a python dictionary.
+///
+/// Returns a named tuple (pyarrow.MonthDayNanoTuple) containing attributes
```
```suggestion
/// Returns a named tuple (pyarrow.MonthDayNano) containing attributes
```

jorisvandenbossche commented on cpp/src/arrow/python/arrow_to_python_internal.h:
```diff
@@ -0,0 +1,52 @@
+// specific language governing permissions and limitations
+// under the License.
+
+// Functions for converting between pandas's NumPy-based data representation
+// and Arrow data structures
```
I think this comment is a left-over from copying another file?

jorisvandenbossche reviewed (commented):
Looks good! Added a few inline comments. And some additional non-inline comments:

- Can you also add it to `test_types.py` (eg adding to `get_all_types()` will automatically run some tests) 
- For consistency with other types, we might also want to add a `is_interval` function in `types.py`
- Add it to `test_type_for_alias` in `test_schema.py`
- Add the type factory to docs in `/python/api/datatypes.rst` and `arrays.rst`

I can also push myself for those (trivial) changes if you want.

One other question (but doesn't need to be handled here):

- Should we allow creating an interval array from plain tuples instead of only from the MonthDayNano named tuple?

pitrou commented:
@emkornfield It can live somewhere with the other Python helpers IMHO.

emkornfield commented:
@jorisvandenbossche thanks for the feedback.  Please see responses inline

> Looks good! Added a few inline comments. And some additional non-inline comments:
> 
> * Can you also add it to `test_types.py` (eg adding to `get_all_types()` will automatically run some tests)

It looks like this has been refactored.  Please let me know if I did it correctly.


> 
> I can also push myself for those (trivial) changes if you want.
> 

I think I've addressed all the feedback here, let me know if I've missed something.

> One other question (but doesn't need to be handled here):
> 
> * Should we allow creating an interval array from plain tuples instead of only from the MonthDayNano named tuple?

Possibly, I think we can add this if users request it later.  I don't think there is a need to do it here because tuples already get inferred as lists.
